FUL-24035: Handle celery worker crash

aga774 opened the pull request:
Changed the default behaviour to handle celery worker crash. Allow the message to be rejected and requeued so it will be executed again by another worker.

aga774 committed:
FUL-24035:
Changed the default behaviour to handle celery worker crash. Allow the message to be rejected and requeued so it will be executed again by another worker.

annaskiba commented on celery/worker/job.py:
```diff
@@ -437,8 +437,14 @@ def on_failure(self, exc_info):
+                #: For acks_late set to True, we changed the default behavior
+                #: to handle worker crash. Currently, we allow the message to be rejected
+                #: and requeued so it will be executed again by another worker.
+                self.reject(requeue=True)
```
I think the original idea was not to requeue it, but to just nack it, @grappeq can you confirm? 
Aga, have you checked that if the task is not requeued but just nacked then the import does not get stuck?

aga774 replied:
I checked only the behaviour with reject, in which a  import has a chance to finish successfully. In all my cases that I have tested, crashed import finished successfully when I was rejecting them and requeuing.

grappeq replied:
I think we should requeue, not nack. Afaik nacking means that the task will never be executed which e.g. for user imports will cause the import to get stuck. On the other hand we are fine with requeuing as `acks_late` tasks either way have to be idempotent so it being executed twice is an expected scenario.

annaskiba replied:
Makes sense, I think the code in other request.py would actually nack (unless we set some flag), thus my question. But I agree with requeuing

annaskiba commented on celery/worker/job.py:
```diff
@@ -437,8 +437,14 @@ def on_failure(self, exc_info):
+
             # (acks_late) acknowledge after result stored.
-            if self.task.acks_late:
+            if self.task.acks_late and isinstance(exc, WorkerLostError):
```
I'd change the ifs to this:
```
   if acks.late:
      if isinstance(exc, WorkerLostError):
         ...
      else:
         ack()
```
I find it more readable, wdyt?

annaskiba commented on celery/worker/job.py:
```diff
@@ -437,8 +437,14 @@ def on_failure(self, exc_info):
+            if self.task.acks_late and isinstance(exc, WorkerLostError):
+                #: For acks_late set to True, we changed the default behavior
+                #: to handle worker crash. Currently, we allow the message to be rejected
+                #: and requeued so it will be executed again by another worker.
```
@grappeq do we want to change the behavior for all our tasks or do we prefer to add the property that will be a flag (like original reject_on_worker_lost) whether this should be used (what should be a default of the flag is another discussion)

grappeq replied:
I would change this for all acks_late tasks as I cannot think of scenario where keeping the default behavior is preferable. Also, other tech leads are aware of the change we are making and they have not requested keeping this behavior.

annaskiba replied:
ok, I wanted to confirm that we're that brave :D

grappeq reviewed (approved):
Looks good to me :smile: The only remaining topic is the one we discussed in person so handling redelivered messages to avoid requeuing loop.

grappeq reviewed (changes requested):
I think you also need to update the version of the package. From what I can see it can be done in this file: https://github.com/beekpr/celery/blob/master/celery/__init__.py#L22

aga774 committed:
FUL-24035:
Upgrade version. Handling redelivered messages to avoid requeuing loop.

annaskiba commented on celery/__init__.py:
```diff
@@ -19,7 +19,7 @@
 
 SERIES = 'Cipater'
-VERSION = version_info_t(3, 1, 25, '', '')
+VERSION = version_info_t(3, 1, 26, '', '')
```
I think this will be confusing - it's not actually a 3.1.26 version as in original celery but our own version. On the other hand, I understand that you'll release it with some other version to be consumed by the monolith. I'm not sure how to handle well such a case.

aga774 replied:
changed to 3.1.25-beekeeper-1 version

annaskiba reviewed (commented):
> I think you also need to update the version of the package. From what I can see it can be done in this file: https://github.com/beekpr/celery/blob/master/celery/__init__.py#L22

See my comment there, I'm not sure if we're not making it confusing this way, but I'm also not sure how to make it properly.

aga774 committed:
FUL-24035:
Change version to indicate that it is a Beekeeper fork of celery in version 3.1.25

grappeq reviewed (approved):
Looks good! Let's test it out :smile:
